contracts/Kernel.sol

```diff
@@ -155,81 +128,203 @@ contract Kernel is Factory {
+                        mstore(0x40,add(result,rsize))
+                    }
+
+                    // Retrieve the address of new available memory from address 0x40
+                    // we will use this as the start of the input (ins)
+                    let ins
+                    let inl
+                    if dataLength {
+                        // If there is any data associated with this procedure
+                        // call (this inlcudes the data such as a function
+                        // selector) we need to set that as the input data to
+                        // the delegatecall.
+                        // First we must allocate some memory.
+                        ins :=  malloc(dataLength)
+                        // Then we store that data at this allocated memory
+                        // location
+                        calldatacopy(ins, 97, dataLength)
+                        inl := dataLength
+                    }
+                    if iszero(dataLength) {
+                        // If there is not data to be sent we just set the
+                        // location and length of the input data to zero. The
+                        // location doesn't actually matter as long the length
+                        // is zero.
+                        ins := 0
+                        inl := 0
+                    }
+                    let retLoc := mallocZero(returnLength)
+                    let status := delegatecall(
+                        // The gas we are budgeting, which is always all the
+                        // available gas
+                        gas,
+                        // The address for the chosen procedure which we
+                        // obtained earlier
+                        procedureAddress,
+                        // The starting memory offset of the innput data
+                        ins,
+                        // The length of the input data
+                        inl,
+                        // The starting memory offset to place the output data
+                        retLoc,
+                        // The length of the output data
+                        returnLength)
+                    // We need to restore the previous procedure as the current
+                    // procedure, this can simply be on the stack
+                    sstore(currentProcedure_slot,div(previousProcedure,exp(0x100,8)))
+                    if iszero(status) {
+                        let errStore := malloc(0x20)
+                        mstore(errStore,add(22,mload(retLoc)))
+                        revert(errStore,0x20)
+                    }
+                    if eq(status,1) {
+                        return(retLoc,returnLength)
+                    }
+                }
+            } else {
+                assembly {
+                    // 33 means the capability was rejected
+                    mstore(0,33)
+                    revert(0,0x20)
```

Comment:
To me a while to realize that revert here undo's line `currentProcedure = procedureKey;`. A comment that mention's this would suffice.

Reply:
Are you sure `mstore` isn't reverted?

Reply:
Everything is reverted, as if it never happened, as if the syscall was never attempted. Which is what we want.

Reply:
If that is the case is the `mstore(0,33)` not returning an error code?

Reply:
Revert still returns that data. Memory is always dropped after a contract's execution with or without revert. So yes the error code is returned.